iree/turbine/kernel/wave/memory_analysis/minimize_shared_allocs.py

```diff
@@ -0,0 +1,187 @@
+        custom = get_custom(node)
+        if isinstance(custom, NestedRegionOp):
+            update_sort_keys(
+                trace,
+                trace.region_graph.subgraphs[custom.subgraph_name],
+                node._sort_key,
+            )
+
+
+@dataclass
+class LiveInterval:
+    start: tuple[int] = (10000,)
+    end: tuple[int] = (-1,)
+
+
+def compute_live_intervals(allocs: list[fx.Node]):
+    """
+    Compute the live intervals for the allocs.
+    """
+    live_intervals = {}
+    for alloc in allocs:
+        live_intervals[alloc] = LiveInterval()
+        users, _ = get_users(alloc, None)
+        for user in users:
+            if user._sort_key < live_intervals[alloc].start:
+                live_intervals[alloc].start = user._sort_key
+            if user._sort_key > live_intervals[alloc].end:
+                live_intervals[alloc].end = user._sort_key
+    return live_intervals
+
+
+def get_shared_memory_allocation_size(alloc: fx.Node, dtype: DataType) -> int:
+    custom = get_custom(alloc)
+    return math.prod([subs_idxc(x) for x in custom.distributed_shape]) * (
+        dtype.bitwidth() // 8
+    )
+
+
+def get_use(
+    alloc: fx.Node, live_interval: LiveInterval, match_sort_key: int
+) -> fx.Node:
+    users, _ = get_users(alloc, None)
+    matches = [x for x in users if x._sort_key == live_interval.start]
+    if len(matches) != 1:
+        raise ValueError(
+            f"Expected 1 match for {alloc} and {match_sort_key}, got {len(matches)}"
+        )
+    return matches[0]
+
+
+def get_first_use(alloc: fx.Node, live_interval: LiveInterval) -> fx.Node:
+    return get_use(alloc, live_interval, live_interval.start)
+
+
+def get_last_use(alloc: fx.Node, live_interval: LiveInterval) -> fx.Node:
+    return get_use(alloc, live_interval, live_interval.end)
+
+
+def insert_barrier_if_needed(alloc: fx.Node, first_use: fx.Node, last_use: fx.Node):
+    """
+    This function inserts a barrier between the last use of the allocation i
+    and the first use of allocation j, if no barrier already exists between them.
+    The barrier is inserted before the first use of allocation j, unless
+    allocation i and allocation j are in different graphs in which case
+    the barrier is inserted before the first use parent op.
+    """
+    if is_barrier_between(last_use, first_use):
+        return
+
+    if last_use.graph != first_use.graph:
+        first_use = first_use.graph.parent_op
+    with first_use.graph.inserting_before(first_use):
+        SharedMemoryBarrier().add_to_graph(first_use.graph)
+
+
+def minimize_shared_allocs(trace: CapturedTrace, minimize_shared_allocs: bool):
+    """
+    Minimize the number of shared allocs by reusing them.
+    See: iree/turbine/docs/kernel/memory_analysis.md for more details.
+    """
+    if not minimize_shared_allocs:
+        return
+    update_sort_keys(trace, trace.get_root_graph())
+
+    allocs = trace.walk(lambda x: isinstance(get_custom(x), Allocate))
```

Comment:
Probably better to check `address_space==SHARED_ADDRESS_SPACE` too, even if it can't be anything else currently.

Reply:
okay will add in subsequent PR.